std/algorithm/comparison.d

```diff
@@ -1911,3 +1915,101 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
         [mytuple(2, 3), mytuple(1, 2)]
     ));
 }
+
+enum ifTestable(T, alias pred = a => a) = __traits(compiles, { if ((pred(T()))) {} }) || isArray!T;
```

Comment:
You can partially apply `pred` to pass it to `allSatisfy` with something like this:
```d
enum ifTestable(alias pred, T) = __traits(compiles, { if ((pred(T.init))) {} }) || isArray!T;

template applyLeft(alias higher, appliedArgs...)
{
   enum applyLeft(args...) = higher!(appliedArgs, args);
}

CommonType!Ts either(alias pred = a => a, Ts...)(lazy Ts a)
    if (a.length > 1 &&
        !is(CommonType!Ts == void) &&
        allSatisfy!(applyLeft!(ifTestable, unaryFun!pred), Ts))
{
    ...
}